### ISSUE TITLE
Possible Race Condition: request.body() is empty after accessing request.queryParams()

### ISSUE THREAD

**Issue by dpx-christian** [perwendel/spark#867]
Hi,

I am not sure if it an issue with jetty or spark. However, when I access request.queryParams() first and request.body() second I have a completely reproducible issue.

The issue is as follow in code:
`		req.queryParams().forEach((v) -> System.out.println(v));
		System.out.println("BODY = " + req.body());
`
If I comment out the first line (req.queryParams()...) it works as expected and req.body() returns the body of the request.

For analysis: I could hunt it down to method Request.readBodyAsBytes() where the method getInputStream() on the servletRequest variable return zero bytes!

Best, Christian

**Comment by dpx-christian** [comment]
oh and btw I checked with the most recent version of spark (2.6.0) and jetty (9.4.5, also with 9.4.4 since is shipped with spark 2.6.0).

/Christian

**Comment by mlitcher** [comment]
I think this is the same as issue #666.

**Comment by dpx-christian** [comment]
@mlitcher 
You are totally right! After having read this "old" bug description I checked back with my code... IFF I switch "Content-Type" from default (application/x-www-form-urlencoded) to e.g. "application/json" it - all of a sudden - works!

My 50 cents: Ok, it _might_ be ignorant/inappropriate HTTP to do as I did, BUT... for any developer out there this first of all looks like a serious BUG! Spark et al have the role of an API like behaviour to a web server like jetty and it simply cannot be that I get an empty body and have no further hints what went wrong.

IMHO there has to be a better handling of such "mistakes" by the user (developer). I would assume those things are kindly handled by frameworks like Spark.

Best, Christian

**Comment by pgoergler** [comment]
Same issue here.
I would like to add "before" filter to log raw request headers & body.
But if it try to read body(), then queryParams became empty ...

**Comment by michalkowol** [comment]
Is there any chance that issue will be fixed soon? I have the same problem.

**Comment by doubler84** [comment]
Hello,
I use spark 2.8.0 and I have the same issue.

`Spark.before((request, response) -> { RestfulMockLog log = new RestfulMockLog("LOG", request.ip(), request.body()); });`

If i try to do **request.body()** , the result is always empty. It's a bug, anyone could help me, I'm using spark also in production enviroment
